source/dub/dependencyresolver.d

```diff
@@ -54,236 +54,149 @@ class DependencyResolver(CONFIGS, CONFIG) {
-
-				// get the current config/version of the current dependency
-				sizediff_t childidx = package_indices[basepack];
-				auto child = configs[childidx];
-
-				if (child.allConfigs.length == 1 && child.allConfigs[0] == CONFIG.invalid) {
-					// ignore invalid optional dependencies
-					if (ch.depType != DependencyType.required)
-						continue;
-
-					if (parentbase == root_base_pack) {
-						import std.uni : toLower;
-						auto lp = ch.pack.toLower();
-						if (lp != ch.pack) {
-							logError("Dependency \"%s\" of %s contains upper case letters, but must be lower case.", ch.pack, parent.pack);
-							if (getAllConfigs(lp).length) logError("Did you mean \"%s\"?", lp);
-						}
-						if (child.anyConfig)
-							throw new Exception(format("Root package %s reference %s %s cannot be satisfied.\nPackages causing the conflict:\n\t%s",
-								parent.pack, ch.pack, ch.configs,
-								child.origins.map!(a => a.origin.pack ~ " depends on " ~ a.dependency.configs.to!string).join("\n\t")));
-						else
-							throw new Exception(format("Root package %s references unknown package %s", parent.pack, ch.pack));
-					}
-					// choose another parent config to avoid the invalid child
-					if (parentidx > maxcpi) {
-						error = ConflictError(ConflictError.Kind.invalidDependency, parent, ch, CONFIG.invalid);
-						logDiagnostic("%s (ci=%s)", error, parentidx);
-						maxcpi = parentidx;
-					}
-				} else {
-					auto config = child.allConfigs[config_indices[childidx]];
-					auto chnode = TreeNode(ch.pack, config);
-
-					if (config == CONFIG.invalid || !matches(ch.configs, config)) {
-						// ignore missing optional dependencies
-						if (config == CONFIG.invalid && ch.depType != DependencyType.required)
-							continue;
-
-						// if we are at the root level, we can safely skip the maxcpi computation and instead choose another childidx config
-						if (parentbase == root_base_pack) {
-							error = ConflictError(ConflictError.Kind.noRootMatch, parent, ch, config);
-							return childidx;
-						}
-
-						if (childidx > maxcpi) {
-							maxcpi = max(childidx, parentidx);
-							error = ConflictError(ConflictError.Kind.childMismatch, parent, ch, config);
-							logDebug("%s (ci=%s)", error, maxcpi);
-						}
-
-						// we know that either the child or the parent needs to be switched
-						// to another configuration, no need to continue with other children
-						if (config == CONFIG.invalid) break;
-					}
-
-					maxcpi = max(maxcpi, validateConfigs(chnode, error));
+			// restrain the configurations to the current dependency spec
+			bool any_config = false;
+			foreach (i, c; di.configs)
+				if (di.included[i]) {
+					if (!matches(dep.configs, c))
+						di.included[i] = false;
+					else any_config = true;
 				}
+
+			if (!any_config && dep.depType == DependencyType.required) {
+				if (di.configs.length)
+					throw new ResolveException(n, dep);
+				else throw new DependencyLoadException(n, dep);
```

Comment:
this formatting seems a little weird (first throw in a new line then in the same line)